---
title: "Modeling trans-eQTL effect using public data"
author: 
  - name: Chris Dong
    affiliation:
    - Department of Statistics and Data Science, University of California, Los Angeles
    email: cycd@g.ucla.edu
  - name: Yihui Cen
    affiliation:
    - Department of Computational Medicine, University of California, Los Angeles
    email: yihuicen@g.ucla.edu
output: 
  BiocStyle::html_document:
    self_contained: yes
    toc: true
    toc_float: true
    toc_depth: 2
    code_folding: show
date: "`r doc_date()`"
package: "`r pkg_ver('scDesignPop')`"
vignette: >
  %\VignetteIndexEntry{scDesignPop-trans-eqtl}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---
```{css, echo=FALSE}
pre {
  white-space: pre !important;
  overflow-x: scroll !important;
}
```

```{r setup, include = FALSE}
knitr::opts_chunk$set(
    message = FALSE,
    collapse = TRUE,
    comment = "#>",
    crop = NULL ## Related to https://stat.ethz.ch/pipermail/bioc-devel/2020-April/016656.html
)
tools::R_user_dir("scDesignPop", which="cache")
```

```{r, message=FALSE, warning=FALSE, results='hide'}
library(scDesignPop)
library(SingleCellExperiment)
library(SummarizedExperiment)
library(ggplot2)
theme_set(theme_bw())
```

## Introduction
scDesignPop supports user-defined eQTL pairs, making it straightforward to extend the framework to trans-eQTL studies with known ground truth.

## Library and data preparation

### prepare the eQTL genotype dataframe as input

Here, we take the trans-eQTL list from eQTLGen phase 1 (https://www.eqtlgen.org/trans-eqtls.html) as an example to show the scDesignPop's flexibility in eQTL ground truth settings. The download link of the trans-eQTL list is https://download.gcc.rug.nl/downloads/eqtlgen/trans-eqtl/2018-09-04-trans-eQTLsFDR0.05-CohortInfoRemoved-BonferroniAdded.txt.gz. 

```{bash}
wget https://download.gcc.rug.nl/downloads/eqtlgen/trans-eqtl/2018-09-04-trans-eQTLsFDR0.05-CohortInfoRemoved-BonferroniAdded.txt.gz
```


Then, we reformat this trans-eQTL list into the required eQTL dataframe as follows,

```{r}
library(scDesignPop)
library(SingleCellExperiment)
library(SummarizedExperiment)
library(ggplot2)
library(data.table)

transeqtl <- fread("2018-09-04-trans-eQTLsFDR0.05-CohortInfoRemoved-BonferroniAdded.txt.gz")
transeqtl <- transeqtl[,c("Gene","SNP","SNPChr","SNPPos")]
transeqtl <- cbind("bin", transeqtl)
colnames(transeqtl) <- c("cell_type","gene_id","snp_id","CHR","POS")
transeqtl <- transeqtl[!duplicated(transeqtl$gene_id), ]
transeqtl$snp_id <- paste(transeqtl$CHR,transeqtl$POS,sep = ":")
```

Since the genotype data are not provided and are often under restricted access, here we extract the SNPs that we have genotype data in the `example_eqtlgeno`

```{r}
data("example_eqtlgeno")
transeqtl <- transeqtl[which(transeqtl$snp_id%in%example_eqtlgeno$snp_id),]

geno_cols <- grep("^SAMP", names(example_eqtlgeno), value = TRUE)
geno_by_snp <- example_eqtlgeno[!duplicated(example_eqtlgeno$snp_id), 
                                c("snp_id", geno_cols)]
idx <- match(transeqtl$snp_id, geno_by_snp$snp_id)
trans_eqtlgeno <- cbind(
  transeqtl,
  geno_by_snp[idx, geno_cols, drop = FALSE]
)
trans_eqtlgeno
```


## Modeling and simulation

### Step 1: construct a data list

To run scDesignPop, a list of data is required as input. This is done using the `constructDataPop` function. A `SingleCellExperiment` object and an `eqtlgeno` dataframe are the two main inputs needed. The `eqtlgeno` dataframe consists of eQTL annotations (it must have cell state, gene, SNP, chromosome, and position columns at a minimum), and genotypes across individuals (columns) for every SNP (rows). Here, we use `example_sce` to filter for overlapped genes in `trans_eqtlgeno' object created


```{r}
data("example_sce")
overlap_genes <- intersect(rownames(example_sce),trans_eqtlgeno$gene_id)
trans_eqtlgeno <- trans_eqtlgeno[which(trans_eqtlgeno$gene_id%in%overlap_genes),]
example_sce <- example_sce[which(rownames(example_sce)%in%overlap_genes),]
```

```{r}
data_list <- constructDataPop(
    sce = example_sce,
    eqtlgeno_df = trans_eqtlgeno,
    new_covariate = as.data.frame(colData(example_sce)),
    overlap_features = NULL,
    sampid_vec = NULL,
    copula_variable = "cell_type",
    slot_name = "counts",
    snp_mode = "single",
    celltype_colname = "cell_type",
    feature_colname = "gene_id",
    snp_colname = "snp_id",
    loc_colname = "POS",
    chrom_colname = "CHR",
    indiv_colname = "indiv",
    prune_thres = 0.9
    )
```

### Step 2: fit marginal model

Next, a marginal model is specified to fit each gene using the
`fitMarginalPop` function.  
Here we use a Negative Binominal as the parametric model using `"nb"`.

```{r}
marginal_list <- fitMarginalPop(
    data_list = data_list,
    mean_formula = "(1|indiv) + cell_type",
    model_family = "nb",
    interact_colnames = "cell_type",
    parallelization = "pbmcapply",
    n_threads = 20L,
    loc_colname = "POS",
    snp_colname = "snp_id",
    celltype_colname = "cell_type",
    indiv_colname = "indiv",
    filter_snps = TRUE,
    snpvar_thres = 0,
    force_formula = FALSE,
    data_maxsize = 1
    )
```

### Step 3: fit a Gaussian copula

The third step is to fit a Gaussian copula using the `fitCopulaPop`
function.

```{r}
set.seed(123, kind = "L'Ecuyer-CMRG")

copula_fit <- fitCopulaPop(
    sce = example_sce,
    assay_use = "counts",
    input_data = data_list[["new_covariate"]],
    marginal_list = marginal_list,
    family_use = "nb",
    copula = "gaussian",
    n_cores = 2L,
    parallelization = "mcmapply"
    )

RNGkind("Mersenne-Twister")  # reset
```

### Step 4: extract parameters

The fourth step is to compute the mean, sigma, and zero probability
parameters using the `extractParaPop` function.

```{r}
para_new <- extractParaPop(
    sce = example_sce,
    assay_use = "counts",
    marginal_list = marginal_list,
    n_cores = 2L,
    family_use = "nb",
    indiv_colname = "indiv",
    new_covariate = data_list[["new_covariate"]],
    new_eqtl_geno_list = data_list[["eqtl_geno_list"]],
    data = data_list[["covariate"]],
    parallelization = "pbmcmapply"
    )
```

### Step 5: simulate counts

The fifth step is to simulate counts using the `simuNewPop` function.

```{r}
set.seed(123)

newcount_mat <- simuNewPop(
    sce = example_sce,
    mean_mat = para_new[["mean_mat"]],
    sigma_mat = para_new[["sigma_mat"]],
    zero_mat = para_new[["zero_mat"]],
    quantile_mat = NULL,
    copula_list = copula_fit[["copula_list"]],
    n_cores = 2L,
    family_use = "nb",
    nonnegative = TRUE,
    input_data = data_list[["covariate"]],
    new_covariate = data_list[["new_covariate"]],
    important_feature = copula_fit[["important_feature"]],
    filtered_gene = data_list[["filtered_gene"]],
    parallelization = "pbmcmapply"
    )
```

### Step 6: create SingleCellExperiment object using simulated data

After simulating the data, we can create a `SingleCellExperiment` object
as follows.

```{r}
simu_sce <- SingleCellExperiment(list(counts = newcount_mat), 
                                 colData = data_list[["new_covariate"]])
names(assays(simu_sce)) <- "counts"

rowData(simu_sce) <- rowData(example_sce)
simu_sce
```

## Session information
```{r}
sessionInfo()
```
